### PR TITLE
Add request IDs to StripeExceptions.

### PR DESCRIPTION
**Pull request by cupcait**

### CODE REVIEW

**Comment by cupcait** [comment]
r? @kyleconroy 
cc @jimdanz

File: src/main/java/com/stripe/exception/StripeException.java [hunk 1]
```diff
@@ -2,14 +2,32 @@
...
 
-	public StripeException(String message, Throwable e) {
+	public StripeException(String message, String requestId, Throwable e) {
 		super(message, e);
+		this.requestId = requestId;
 	}
 
 	private static final long serialVersionUID = 1L;
 
+    public String getRequestId() {
```
Comment (kyleconroy) [inline src/main/java/com/stripe/exception/StripeException.java]: Looks like the spacing got messed up here.

File: src/main/java/com/stripe/exception/StripeException.java [hunk 2]
```diff
@@ -2,14 +2,32 @@
...
+		return requestId;
+    }
+
+    public String toString() {
+		String reqIdStr = "";
+		if (requestId != null) {
+			reqIdStr = "; request-id: " + requestId;
+        }
+
```
Comment (kyleconroy) [inline src/main/java/com/stripe/exception/StripeException.java]: nit: no need for this empty line

**Comment by kyleconroy** [comment]
We should add a test that checks the request ID can be accessed from an error. `testChargeRetrieveNullId` seems like the best place to put this.

**Comment by kyleconroy** [comment]
@jimdanz any opinion on adding http response headers to exceptions? Since we don't add the body or the status code, we decided just adding the request ID would be better.

File: src/main/java/com/stripe/net/LiveStripeResponseGetter.java [hunk 3]
```diff
@@ -502,20 +510,20 @@ private static StripeResponse getMultipartStripeResponse(
...
 		case 404:
-			throw new InvalidRequestException(error.message, error.param, null);
+			throw new InvalidRequestException(error.message, requestId, error.param, null);
 		case 401:
 			throw new AuthenticationException(error.message);
 		case 402:
-			throw new CardException(error.message, error.code, error.param, error.decline_code, error.charge, null);
+			throw new CardException(error.message, requestId, error.code, error.param, error.decline_code, error.charge, null);
 		default:
 			throw new APIException(error.message, null);
```
Comment (jimdanz) [inline src/main/java/com/stripe/net/LiveStripeResponseGetter.java]: Do we never return a request ID on 500s?  We definitely record the iar in some cases.

**Comment by jimdanz** [comment]
@kyleconroy I'm +1 to just the request id!

File: src/main/java/com/stripe/net/LiveStripeResponseGetter.java (hunk 3 above)
Comment (kyleconroy) [inline src/main/java/com/stripe/net/LiveStripeResponseGetter.java]: Yep, the requestId should be added in all these cases.

**Comment by cupcait** [comment]
@kyleconroy @jimdanz updated! Can one of you take another look? Thanks!

File: src/test/java/com/stripe/StripeTest.java [hunk 4]
```diff
@@ -344,7 +344,9 @@ public void testChargeRetrieveNullId() throws StripeException {
 			assertTrue(false);
 		}
 		catch (InvalidRequestException e) {
-			// Expected
+			String requestId = e.getRequestId();
+			assertTrue(requestId != null);
```
Comment (kyleconroy) [inline src/test/java/com/stripe/StripeTest.java]: Can we check that it also isn't an empty string?

File: src/main/java/com/stripe/net/LiveStripeResponseGetter.java [hunk 5]
```diff
@@ -363,8 +364,16 @@ private static StripeResponse makeURLConnectionRequest(
 			}
 			int rCode = response.responseCode;
 			String rBody = response.responseBody;
+
+			String requestId = "";
+			Map<String, List<String>> headers = response.getResponseHeaders();
```
Comment (kyleconroy) [inline src/main/java/com/stripe/net/LiveStripeResponseGetter.java]: I believe that headers can be null here.

**Comment by kyleconroy** [comment]
Other than my two notes, looks good.

**Comment by kyleconroy** [comment]
:+1:

**Pull request merged**